packages/node/src/login/oidc/refresh/TokenRefresher.ts

```diff
@@ -42,21 +50,65 @@ export interface ITokenRefresher {
 export default class TokenRefresher implements ITokenRefresher {
   constructor(
     @inject("storageUtility") private storageUtility: IStorageUtility,
-    @inject("tokenRequester") private tokenRequester: ITokenRequester
+    @inject("issuerConfigFetcher")
+    private issuerConfigFetcher: IIssuerConfigFetcher,
+    @inject("clientRegistrar") private clientRegistrar: IClientRegistrar
   ) {}
 
-  async refresh(localUserId: string): Promise<void> {
-    // Get the refresh token and the issuer
-    const refreshToken = await this.storageUtility.getForUser(
-      localUserId,
-      "refreshToken",
-      { errorIfNull: true, secure: true }
+  async refresh(
+    sessionId: string,
+    refreshToken?: string,
+    dpopKey?: JWK.ECKey
+  ): Promise<TokenSet> {
+    const oidcContext = await loadOidcContextFromStorage(
+      sessionId,
+      this.storageUtility,
+      this.issuerConfigFetcher
     );
-    /* eslint-disable camelcase */
-    await this.tokenRequester.request(localUserId, {
-      grant_type: "refresh_token",
-      refresh_token: refreshToken as string, // This cast can be safely made because getForUser will error if refreshToken is null
+    const issuer = new Issuer(configToIssuerMetadata(oidcContext.issuerConfig));
+    // This should also retrieve the client from storage
+    const clientInfo: IClient = await this.clientRegistrar.getClient(
+      { sessionId },
+      oidcContext.issuerConfig
+    );
+    const client = new issuer.Client({
+      client_id: clientInfo.clientId,
+      client_secret: clientInfo.clientSecret,
+    });
+
+    if (refreshToken === undefined) {
+      // TODO: in a next PR, look up storage for a refresh token
+      throw new Error(
+        `Missing a refresh token to refresh the access token associated to the session [${sessionId}]`
+      );
+    }
+
+    if (oidcContext.dpop && dpopKey === undefined) {
+      throw new Error(
+        `The key bound to the access token associated to the session [${sessionId}] must be provided to refresh it.`
```

Comment:
I'm not sure exactly the best wording here, but the current message text makes no reference to DPoP at all, while the conditional statement is explicitly checking for it, so maybe something like:
```suggestion
        `For session [${sessionId}], the key bound to the access token (i.e. the DPoP key) must be provided to allow us refresh our access token.`
```

Reply:
A key-bound access token is a DPoP token. I'll rephrase.

Reply:
Resolved in 8818797c031a7e51ec3c56e1bf2e195d54ff6dee